lls/cache.c

```diff
@@ -351,39 +350,80 @@ xmit_icmp_ping_reply(struct gatekeeper_if *iface, struct rte_mbuf *pkt)
 			RTE_ETHER_TYPE_IPV4);
 	}
 
-	/*
-	 * Set-up IPv4 header.
-	 *
-	 * According to RFC 792 page 13: to form an echo reply
-	 * message, the source and destination addresses are simply reversed,
-	 * the type code changed to 0, and the checksum recomputed.
-	 */
+	/* Set-up IPv4 header. */
 	icmp_ipv4 = (struct rte_ipv4_hdr *)pkt_out_skip_l2(iface, icmp_eth);
+	icmph = (struct rte_icmp_hdr *)ipv4_skip_exthdr(icmp_ipv4);
 
 	icmp_ipv4->time_to_live = IP_DEFTTL;
-	ip_addr_tmp = icmp_ipv4->src_addr;
-	icmp_ipv4->src_addr = icmp_ipv4->dst_addr;
-	icmp_ipv4->dst_addr = ip_addr_tmp;
+	if (icmph->icmp_type == ICMP_ECHO_REQUEST_TYPE) {
+		/*
+		 * According to RFC 792 page 13: to form an echo reply
+		 * message, the source and destination addresses are simply
+		 * reversed, the type code changed to 0, and the checksum
+		 * recomputed.
+		 */
+		uint32_t ip_addr_tmp = icmp_ipv4->src_addr;
+		icmp_ipv4->src_addr = icmp_ipv4->dst_addr;
+		icmp_ipv4->dst_addr = ip_addr_tmp;
+
+		/*
+		 * According to RFC 792 page 14:
+		 *
+		 * (1) the data received in the echo message must be returned in
+		 * the echo reply message.
+		 *
+		 * (2) the identifier and sequence number may be used by
+		 * the echo sender to aid in matching the replies with
+		 * the echo requests.
+		 *
+		 * So, we keep these fields unmodified.
+		 */
+		icmph->icmp_type = ICMP_ECHO_REPLY_TYPE;
+		icmph->icmp_code = ICMP_ECHO_REPLY_CODE;
+	} else if (likely(icmph->icmp_type == ICMP_MASK_REQUEST_TYPE)) {
+		uint32_t minimum_pkt_size;
+		struct in_addr *ip4_mask;
+
+		/*
+		 * According to RFC 3021: the 255.255.255.255 IP broadcast
+		 * address MUST be used for broadcast Address Mask Replies
+		 * in point-to-point links with 31-bit subnet masks.
+		 * In other cases, the form {<Network-number>, -1} should be
+		 * used.
+		 */
+		icmp_ipv4->src_addr = icmp_ipv4->dst_addr;
+		if (iface->ip4_addr_plen >= 31)
+			icmp_ipv4->dst_addr = (uint32_t)-1;
+		else {
+			icmp_ipv4->dst_addr = (iface->ip4_addr.s_addr |
+					(~iface->ip4_mask.s_addr));
+		}
+
+		minimum_pkt_size = sizeof(*icmp_eth) + sizeof(*icmp_ipv4) +
+			sizeof(*icmph) + sizeof(struct in_addr);
```

Comment:
Notice that we use `ipv4_skip_exthdr()` to obtain the variable `icmph`, so the IPv4 header may have extensions. The Ethernet header too can vary size. A better strategy is to check the sizes in the headers as needed and append the needed size. Please test this code once it's rewritten.

Reply:
Perhaps just accounting for the variable sizes of the Ethernet and IP headers in the expression above is enough.

Reply:
Thanks @AltraMayor ! Just accounted for the variable sizes of both Ethernet and IP headers, but haven't got chance to do the test yet. Will do the tests later.